nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/PublishMQTT.java

```diff
@@ -0,0 +1,288 @@
+            .expressionLanguageSupported(true)
+            .required(true)
+            .addValidator(StandardValidators.createAttributeExpressionLanguageValidator(AttributeExpression.ResultType.STRING, true))
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_QOS = new PropertyDescriptor.Builder()
+            .name("Quality of Service(QoS)")
+            .description("The Quality of Service(QoS) to send the message with. Accepts three values '0', '1' and '2'; '0' for 'at most once', '1' for 'at least once', '2' for 'exactly once'.")
+            .required(true)
+            .expressionLanguageSupported(true)
+            .addValidator(QoSValidator)
+            .build();
+
+    public static final PropertyDescriptor PROP_RETAIN = new PropertyDescriptor.Builder()
+            .name("Retain Message")
+            .description("Whether or not the retain flag should be set on the MQTT message.")
+            .required(true)
+            .expressionLanguageSupported(true)
+            .addValidator(RetainValidator)
+            .build();
+
+    public static final PropertyDescriptor PROP_TIMEOUT = new PropertyDescriptor.Builder()
+            .name("QoS Timeout")
+            .description("How many milliseconds to wait for the broker to acknowledge delivery of a message before routing the message back to the source queue for retry. Only applicable " +
+                    "when QoS is 1 or 2")
+            .required(true)
+            .addValidator(StandardValidators.NON_NEGATIVE_INTEGER_VALIDATOR)
+            .build();
+
+    private List<PropertyDescriptor> descriptors;
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("FlowFiles that are sent successfully to the destination are sent out this relationship.")
+            .build();
+    public static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("FlowFiles that failed to send to the destination are sent out this relationship.")
+            .build();
+
+    private Set<Relationship> relationships;
+    private ProcessorLog logger;
+    private MqttConnectOptions connOpts;
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        logger = getLogger();
+        final List<PropertyDescriptor> descriptors = getAbstractPropertyDescriptors();
+        descriptors.add(PROP_TOPIC);
+        descriptors.add(PROP_QOS);
+        descriptors.add(PROP_RETAIN);
+        descriptors.add(PROP_TIMEOUT);
+        this.descriptors = Collections.unmodifiableList(descriptors);
+
+        final Set<Relationship> relationships = new HashSet<Relationship>();
+        relationships.add(REL_SUCCESS);
+        relationships.add(REL_FAILURE);
+        this.relationships = Collections.unmodifiableSet(relationships);
+    }
```

Comment:
Given NIFI-1384 and NIFI-1318 I would discourage doing the above  in the init method and instead use static initializer.